kvbc/src/sparse_merkle/tree.cpp

```diff
@@ -141,11 +185,17 @@ UpdateBatch Tree::update_impl(const concord::kvbc::SetOfKeyValuePairs& updates,
     sparse_merkle::remove(walker, key_hash);
   }
 
+  // Compute value hashes in parallel with tree updates.
+  // Assume the same order for subsequent iterations of SetOfKeyValuePairs objects, provided that we don't mutate them
+  // in any way. That, combined with the assumption that the tree update takes longer than value hash calculation,
+  // should ensure that this is actually an optimization.
+  // Note: The documentation for std::unordered_map (SetOfKeyValuePairs) states that read-only operations do not
+  // invalidate iterators. Therefore, we can expect that subsequent iterations will have the same order.
+  auto value_hashes = ValueHashes(updates);
   for (auto&& [key, val] : updates) {
-    auto leaf_hash = hasher.hash(val.data(), val.length());
     LeafNode leaf_node{val};
     LeafKey leaf_key{hasher.hash(key.data(), key.length()), version};
-    LeafChild child{leaf_hash, leaf_key};
+    LeafChild child{value_hashes[key], leaf_key};
```

Comment:
Doesn't this block, since get is called on the future?

Reply:
You are correct, it does. Assumption here is that value hash calculation is faster than the tree update. As discussed offline, I am closing this PR for now. Will open another PR that implements an std::async()-like thread pool as an utility. Will further measure performance with it and decide if we should calculate value hashes in parallel and, if yes, how many threads should be used.